### ISSUE TITLE
Disable validation for hidden/nested fields 

### ISSUE THREAD

**Issue by ramiroazar** [sanity-io/sanity#3937]
**Is your feature request related to a problem? Please describe.**

Hiding a field that has nested fields with validation prevents the document from publishing because of validation errors.

Failed validation from more complex hidden fields make the [conditional fields](https://www.sanity.io/docs/conditional-fields) feature a lot less versatile.

```
{
  name: 'template',
  title: 'Template',
  type: 'object',
  hidden: true,
  fields: [
    {
      name: 'title',
      title: 'Title',
      type: 'string',
      validation: Rule => Rule.required()
    },
  ]
}
```

This is intentional according to [this](https://www.sanity.io/schemas/optional-validation-for-hidden-fields-439d0ab8) snippet.

> Validations in the studio start from the premise that the validation should pass regardless of the UI state. For example, a field might be hidden because of an author's role or because other content has not been filled in, but still required for publishing.

However, as also pointed out by the same snippet, if a field is hidden then validation might not be relevant.

I understand that custom validation logic could align with the hidden logic, but this doesn't seem very practical for more advanced scenarios.

What if the hidden field contains custom field types/objects or multiple levels of nesting?

```
// title.ts
{
  name: 'title',
  title: 'Title',
  type: 'object',
  fields: [
    {
      name: 'number',
      title: 'number',
      type: 'number',
      validation: Rule => Rule.required()
    },
    {
      name: 'text',
      title: 'text',
      type: 'string',
      validation: Rule => Rule.required()
    }
  ]
}
// template.ts
{
  name: 'template',
  title: 'Template',
  type: 'object',
  hidden: true,
  fields: [
    {
      name: 'title',
      title: 'title',
    }
  ]
}
```

I appreciate disabling all validation on hidden fields by default might not be the right solution, but it seems like Sanity is at least missing a way of opting in to disabling validation for all nested fields if needed.

**Describe the solution you'd like**

For the `validation` property to disable validation on nested fields when `false`.

Alternatively, a new property similar to the existing `validation` property (ie. boolean/callback) to disable all validation for nested fields.

**Describe alternatives you've considered**

My current use case is grouping a collection of fields based on "templates" and only showing the correct ones based on the "template" selected by the editor from a `string/list` field. 

The main alternative I'm considering is actually refactoring away from hidden object fields and using a `reference` field to select a `document/template` instead.

**Comment by fvieira** [comment]
I strongly agree, having hidden object running validation is annoying, we have an internal rule that any object with `hidden` must have a `validation` that is the exact opposite.
I would love to have a config level option `hiddenFieldDontValidate` to make all hidden fields not run validations by default, and then an option `validateWhenHidden` in fields to override this if needed (and I'm not even sure when I'd ever use this, but it's the current default so maybe someone wants it).

**Comment by fvieira** [comment]
Just want to add that I got hit by this again, and in an even more annoying way, as the validation error is in a deeply nested value of the hidden object, so making the validation there depend on whether the far removed ancestor is hidden is not practical at all, this should be just automatic.

Luckily, this happened to me, and I knew that I had to make the hidden field appear again and fix the validations on it to proceed, but if it was to a client of mine, I'm sure they'd be completely stumped, and I have no way to cleanly ensure this type of issues don't occur.

**Comment by RitaDias** [comment]
👋 
Hi all,
Thank for reporting and sorry for taking so long getting back to you all!

I brought this up with the team internally, I will be tackling a solution which I think will help!
To be clear though, the approach will not to disable the validation to the hidden fields. The logic as to not do it is the following: 
- an editor wants to publish a document, 
- has added something to a field which would run into a validation concern and not allowed them to publish
- the editor hides the field by doing something in the document (something that is part of the normal flow of some documents which we've seen), 
- the content is still there but because we wouldn't run the validation - which means that something that would have failed before now is published because we didn't run the validation.

In this use case, folks end up having things that shouldn't have been allowed to be published, out there without them potentially realising it (because they've been hidden and it's not clear).

The approach we will go with is to make sure that the validation errors are shown still, but that when you click them you can open the hidden fields directly and change them as needed :)

**Comment by fvieira** [comment]
Hi @RitaDias, I can hardly believe we will finally have a solution for this! It's probably one of the most annoying issues we've been dealing with, as it forces us to be 100% sure that the validations only run for visible fields, which is far from trivial. 

The solution you propose is already a major step up from the current state and I appreciate it and the reasons behind the decision.
Still, just want to leave my opinion that I would still like to have a config level option to make all hidden fields not run validations by default, because while I understand the risks, my use cases are 99% that if a field is hidden, it's because it doesn't matter, the frontend won't look at it, and validations are irrelevant.
Most of the times that we hide fields, it's because the user can pick between option A and option B, both of which make some fields specific to the option appear. So if the user fills the fields of option A in an invalid way and then picks option B, it doesn't matter that the option A fields are wrongly filled, the front-end code won't even look at those fields. Forcing the user to clean those errors is not the best UX.

But again, I understand why you're taking this safe route, and while the UX might still be slightly annoying for editors, it is safe indeed, and it's still a whole lot better than what we have now, which leaves editors completely stumped.

**Comment by RitaDias** [comment]
@fvieira 👋 

This is completely valid feedback which I will log and take back for us to eventually return to consider. I do think that this is a good enough first stab at it and we can iterate as we go :) 

(Also, é sempre bom ver alguém de uma companhia baseada em Lisboa e com um nome português out in the wild 😄 )

Once more I'm really sorry this has taken so long, but I so appreciate this feedback and reply!

**Comment by fvieira** [comment]
Great! That's all I ask for, and I do agree the safe approach is how we should start. My suggestion is more dangerous and needs more thought put into it. 

(Ah ah, pensei o mesmo, mas não consegui confirmar se eras mesmo tuga. 😆 🇵🇹 )

**Comment by RitaDias** [comment]
👋 
Just to keep this updated as we've had a bit more exploration on our side.
I've had to pivot the solution because as I explored the approach mentioned above there were quite a few use cases that popped up that could cause some more gnarly UX issues which leads me to believe that to properly fix this we'll need a bit more time and proper focus,

_However,_ I'll do something else which will unblock most of your use cases (hopefully). I've added to the context within the validation method a way for checking if the field is hidden or not. This will then allow you to say: if the field is hidden, for any reason (the current field is hidden or the ancestors are hidden) just skip the validation.

This is our way of offering you ways of going around the validation right now that doesn't change default behaviour for the reasons I've touched above (and others).

Example of what I mean:

```
defineField({
        name: 'templateA',
        title: 'Template A',
        type: 'object',
        description:
          'Hidden unless "Template A" is selected. Has nested required fields that cause validation errors even when hidden.',
        hidden: ({document}) => document?.selectedTemplate !== 'templateA',
        fields: [
          defineField({
            name: 'title',
            title: 'Title',
            type: 'string',
            validation: (rule, context) =>
               // this means that if an ancestor or the field is hidden then the validation will always just pass vs if it's not
              context.hidden ? rule.skip() : rule.required().min(5),
          }),
          …
        ],
      }),
```

for more deeply nested example:

```
defineField({
        name: 'deeplyNested',
        title: 'Deeply Nested Hidden Fields',
        type: 'object',
        description: 'Demonstrates validation issues with deeply nested hidden fields',
        hidden: ({document}) => document?.selectedTemplate === 'none',
        fields: [
          defineField({
            name: 'level1',
            title: 'Level 1',
            type: 'object',
            fields: [
              defineField({
                name: 'level2',
                title: 'Level 2',
                type: 'object',
                fields: [
                  defineField({
                    name: 'level3',
                    title: 'Level 3',
                    type: 'object',
                    fields: [
                      defineField({
                        name: 'deepRequiredField',
                        title: 'Deep Required Field',
                        type: 'string',
                        validation: (rule, context) => {
                          return context?.hidden ? rule.skip() : rule.min(10)
                        },
                      }),
                    ],
                  }),
                ],
              }),
            ],
          }),
        ],
      }),
```

The actual proper fix that doesn't revolve around this context has been added to our backlog with all the findings I have had currently (but the timing is not great due to more exploration / prioritisation needed).

**Comment by fvieira** [comment]
Hi @RitaDias, thank you for the update and I do understand the decision and workaround.

Having that `context.hidden` won't prevent us from having validations running on hidden fields nor help editors unblock themselves and when it happens. However it does make it a lot easier to skip validations on hidden fields without having to duplicate the whole conditional from the hidden field and it will prevent us from introducing bugs by editing the hidden condition and forgetting to update the validation condition accordingly, so yes, the workaround is appreciated. 

I was thinking that I might even be able to create a patch for Sanity to use that `context.hidden` automatically on every validation which would essentially do what I wanted from the start!
